management/cpu_mem_monitor/include/cpu_mem_monitor/cpu_mem_monitor.h

```diff
@@ -0,0 +1,189 @@
+  void AssertCpuStats();
+  void AssertMemStats();
+
+  // Collect usage stats about all the CPUs and memory, calculate
+  // percentages based on the last time this was called. You should
+  // call this in regular intervals for the numbers to make sense over time.
+  int CollectCPUStats();
+  int CollectMemStats();
+
+  // Callback to scan and publish the CPU and Memory stats
+  void PublishStatsCallback(ros::TimerEvent const &te);
+
+ private:
+  struct Load {
+    std::uint64_t total_time,
+                  user_time,
+                  system_time,
+                  idle_time,
+                  nice_time,
+                  io_time,
+                  irq_time,
+                  soft_irq_time,
+                  steal_time,
+                  guest_time,
+                  system_all_time,
+                  idle_all_time;
+
+    double total_percentage,
+           nice_percentage,
+           user_percentage,
+           system_percentage,
+           virt_percentage;
+  };
+
+  config_reader::ConfigReader config_params_;
+  ros::Publisher cpu_state_pub_;            // Cpu stats publisher
+  ros::Publisher mem_state_pub_;            // Memory stats publisher
+  ros::Timer reload_params_timer_;          // Ckeck if parameters were updated
+  ros::Timer stats_timer_;                  // Update stats
+  ros::Timer assert_cpu_load_fault_timer_;  // Check cpu load limits
+  ros::Timer clear_cpu_load_fault_timer_;   // Clear cpu fault timer
+  ros::Timer assert_mem_load_fault_timer_;  // Check memory load limits
+  ros::Timer clear_mem_load_fault_timer_;   // Clear memory fault timer
+  int pub_queue_size_, update_freq_hz_;     // Monitor publishing queue size and frequency
+  struct sysinfo mem_info_;                 // Scope memory info from sysinfo
+
+  unsigned int ncpus_;          // Number of cpu's
+  std::string processor_name_;  // Processor running this monitor (mlp,llp)
+
+  // Store PID values
+  std::string monitor_host_;     // Either mlp, llp, hlp
+  std::map<std::string, int> nodes_pid_;
+
+  // CPU Monitor Parameters
+  bool temp_fault_triggered_;                // Temperature fault triggered
+  Cpu freq_cpus_;
+  double temperature_scale_;                 // Scale of temperature
+  float avg_cpu_load_high_value_;
+  std::vector<Load> load_cpus_;
+  int cpu_avg_load_limit_, cpu_temp_limit_;
+
+  // Memory Monitor Parameters
+  float mem_load_value_, avg_mem_load_high_value_, mem_load_limit_;
+
+
+  // Status messages to publish
+  ff_msgs::CpuStateStamped cpu_state_msg_;
+  ff_msgs::MemStateStamped mem_state_msg_;
+
+  // Asserting faults
+  LoadFaultState load_fault_state_;
+  int assert_load_high_fault_timeout_sec_, clear_load_high_fault_timeout_sec_;
+};
+
+
+int ParseLine(char* line) {
+  // This assumes that a digit will be found and the line ends in " Kb".
+  int i = strlen(line);
+  const char* p = line;
+  for (uint j = 0; ( *p < '0' || *p > '9' ) && j < strlen(line); ++p, ++j)
```

Comment:
I didn't know you could add multiple lines to the for loop!

Reply:
This is going to execute line[i-3] = '\0'; once each time through the for loop, which won't break anything but probably isn't what you want. Also doesn't matter but you could have also stuck to the while loop and done p - line < i

Reply:
totally forgot the ;! That would have been a slick solution, I didn't do that because p is a character, so I introduced a counter.

Reply:
You could remove the `j` variable and simplify the search like this, as strings always end with `'\0'` (in fact, `'\0'` is what `strlen` itself looks for):

```cpp
for (; (*p < '0' || *p > '9') && *p != '\0'; ++p) {}
```

Or use `std::find_if` to get a pointer to the first digit, and have it handle the boundary conditions:

```cpp
const char* p = std::find_if(line, line + i,
                             [] (char c) { return c >= '0' && c <= '9'; });
```